File: java/betterkhmer/src/test/java/com/betterkhmer/BetterKhmerTest.java
package com.betterkhmer;

import java.io.*;
import java.nio.file.*;
import java.util.*;

public class BetterKhmerTest {
    private static final String FIXTURES = System.getProperty("fixtures.dir",
        System.getProperty("user.dir") + "/../../fixtures");

    private static List<String> loadLines(String name) throws IOException {
        Path p = Paths.get(FIXTURES, name);
        return Files.readAllLines(p);
    }

    public static void main(String[] args) throws Exception {
        // basic
        String got = BetterKhmer.normalize("ខ្មែរ");
        if (!got.equals("ខ្មែរ")) {
            throw new AssertionError("basic FAILED: got " + got);
        }
        System.out.println("basic: ok");

        List<String> inputs   = loadLines("input.txt");
        List<String> expected = loadLines("expected.txt");
        if (inputs.size() != expected.size()) throw new AssertionError("length mismatch");

        int failures = 0;
        for (int i = 0; i < inputs.size(); i++) {
            String result = BetterKhmer.normalize(inputs.get(i));
            if (!result.equals(expected.get(i))) {
                failures++;
                if (failures <= 10) {
                    System.err.println("[" + i + "] got " + result + ", want " + expected.get(i));
                }
            }
        }
        if (failures > 0) {
            throw new AssertionError(failures + "/" + inputs.size() + " fixture failures");
        }
        System.out.println("fixtures: " + inputs.size() + " ok");
    }
}

File: kotlin/betterkhmer/src/test/kotlin/com/betterkhmer/BetterKhmerTest.kt
package com.betterkhmer

import java.nio.file.Files
import java.nio.file.Paths

fun main() {
    val fixturesDir = System.getProperty("fixtures.dir",
        System.getProperty("user.dir") + "/../../fixtures")

    val got = BetterKhmer.normalize("ខ្មែរ")
    check(got == "ខ្មែរ") { "basic FAILED: got $got" }
    println("basic: ok")

    val inputs   = Files.readAllLines(Paths.get(fixturesDir, "input.txt"))
    val expected = Files.readAllLines(Paths.get(fixturesDir, "expected.txt"))
    check(inputs.size == expected.size) { "length mismatch" }

    var failures = 0
    for (i in inputs.indices) {
        val result = BetterKhmer.normalize(inputs[i])
        if (result != expected[i]) {
            failures++
            if (failures <= 10) System.err.println("[$i] got $result, want ${expected[i]}")
        }
    }
    check(failures == 0) { "$failures/${inputs.size} fixture failures" }
    println("fixtures: ${inputs.size} ok")
}
